### PR TITLE
[`GPTNeoX`] Fix BC issue with 4.36

### LINKED ISSUE #28316: Pythia regression in transformers==4.36.2 vs transformers==4.30.1
### System Info

Happy New Year all!

- `transformers` version: 4.36.2
- Platform: Linux-5.15.0-1049-aws-x86_64-with-glibc2.31
- Python version: 3.10.12
- Huggingface_hub version: 0.19.4
- Safetensors version: 0.3.1
- Accelerate version: 0.25.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2+cu121 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): 0.6.8 (cpu)
- Jax version: 0.4.8
- JaxLib version: 0.4.7
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes, via `accelerate`

### Who can help?

Maybe @younesbelkada @ArthurZucker?

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Here is a minimal reproduction https://gist.github.com/vwxyzjn/e67e0bb28363e6fbb309bd0b78922a93. I ran the same `repro.py` with `transformers==4.36.2` and `transformers==4.30.1`, resulting in slightly different losses. Given the data is and other dependencies are precisely the same.

```
python repro.py # 4.36.2
epoch: 0
update: 9, loss: 0.6855486035346985
update: 17, loss: 0.6901922225952148
update: 25, loss: 0.6883461475372314
update: 33, loss: 0.6975809931755066
update: 41, loss: 0.6995139122009277
update: 49, loss: 0.6912401914596558
update: 57, loss: 0.698995053768158
update: 65, loss: 0.7005056142807007
update: 73, loss: 0.7048475742340088
update: 81, loss: 0.6950501203536987
update: 89, loss: 0.7148610949516296
update: 97, loss: 0.694938063621521
update: 105, loss: 0.6957464814186096
update: 113, loss: 0.6873601675033569

python repro.py # 4.30.1
epoch: 0
update: 9, loss: 0.6904680132865906
update: 17, loss: 0.6958459615707397
update: 25, loss: 0.6878675818443298
update: 33, loss: 0.6945885419845581
update: 41, loss: 0.6920362710952759
update: 49, loss: 0.6866860389709473
update: 57, loss: 0.685932457447052
update: 65, loss: 0.6930047273635864
update: 73, loss: 0.6854068636894226
update: 81, loss: 0.6739884614944458
update: 89, loss: 0.6913299560546875
update: 97, loss: 0.7025052309036255
```

# Regression in end-to-end reward model training performance

This difference causes a regression in training reward models. When setting the code, data to be **exactly** the same, the average reward model accuracy across four random seeds is as follows:

* transformers==4.36.2, accelerate==0.25.0, deepspeed==0.12.6
    * EleutherAI/pythia-1b-deduped: 0.6276
    * EleutherAI/pythia-2.8b-deduped: 0.6438 
    * EleutherAI/pythia-6.9b-deduped: 0.65
* transformers==4.30.1, accelerate==0.25.0, deepspeed==0.12.6
    * EleutherAI/pythia-1b-deduped: 0.6327
    * EleutherAI/pythia-2.8b-deduped: 0.6713
    * EleutherAI/pythia-6.9b-deduped: 0.6923

The SFT losses are relatively similar (maybe except for 6.9B, there was a minor loss explosion with `transformers==4.36.2`)

Here is the report. https://wandb.ai/costa-huang/tldr_summarize/reports/pythia-transformers-regression--Vmlldzo2Mzk3OTQ1

image

image

image

Here is the code comparison: identical code and only the dependencies are different
image

image

### Expected behavior

There shouldn't be a regression in the performance.

### LINKED ISSUE #28360: Pythia (GPTNeoXForCausalLM) Regression (inference time) in transformers 4.35.0
### System Info

- `transformers` version: 4.35.0
- Platform: Linux-5.16.19-76051619-generic-x86_64-with-glibc2.35
- Python version: 3.10.11
- Huggingface_hub version: 0.17.3
- Safetensors version: 0.3.1
- Accelerate version: 0.25.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.3.0.dev20240104 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### The Problem

The Pythia models, and by extension the `GPTNeoXForCausalLM` implementation don't appear to be working correctly in 4.35. I've attached a simple reproduction snippet below. This code works on 4.34, but produces NaNs on 4.35 during the forward pass. The token ids are not particularly anomalous.

The problem is likely related to the report at https://github.com/huggingface/transformers/issues/28316, but this issue shows that any effects on reward modeling might be 2nd-order effects and that changes between 4.34 and 4.35 are the problem.

@ArthurZucker @younesbelkada

### Reproduction

```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

model_name_or_path = "EleutherAI/pythia-70m"
model = AutoModelForCausalLM.from_pretrained(model_name_or_path, torch_dtype=torch.float16)


input_ids = [29, 93, 303, 64, 5478, 49651, 10394, 187, 34, 12939, 875]  # this is '<|im_start|>system\nA chat between' 
# alternative: tokenizer('<|im_start|>system\nA chat between')
input_ids = torch.as_tensor(input_ids)[None]

model.cuda()
input_ids = input_ids.cuda()
model(input_ids)["logits"]  # has NaNs?
```

### Expected behavior

Normal forward pass, without NaNs.

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
We broke some of the GPTNeoX model with the dtype in #25830. Sorry for the inconvenience. 
This was breaking in terms of logits, now this PR will probably make the model slow than with casting to smaller dtype.

Fixes #28360, fixes #28316.

For a sample generation I am getting some slow down.

### CODE REVIEW

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing! 

Just for reference and tracking - could you add the number for the generation times before and after this change?

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -549,24 +549,24 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
...
     def forward(self, x, seq_len=None):
         # x: [bs, num_attention_heads, seq_len, head_size]
         if seq_len > self.max_seq_len_cached:
             self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
 
         return (
-            self.cos_cached[:seq_len].to(dtype=x.dtype),
-            self.sin_cached[:seq_len].to(dtype=x.dtype),
+            self.cos_cached[:seq_len].to(x.device),
+            self.sin_cached[:seq_len].to(x.device),
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:562]:
```suggestion
            self.cos_cached[:seq_len],
            self.sin_cached[:seq_len],
```

File: src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py [hunk 2]
```diff
@@ -256,17 +257,17 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
...
     def forward(self, x, seq_len=None):
         # x: [bs, num_attention_heads, seq_len, head_size]
         if seq_len > self.max_seq_len_cached:
             self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
 
         return (
-            self.cos_cached[:seq_len].to(dtype=x.dtype),
-            self.sin_cached[:seq_len].to(dtype=x.dtype),
+            self.cos_cached[:seq_len].to(x.device),
+            self.sin_cached[:seq_len].to(x.device),
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:270]:
```suggestion
            self.cos_cached[:seq_len],
            self.sin_cached[:seq_len],
```
